File: aoc-day3-p2.js
// Put csv input here:
let input = '';
// Stop numbers from being at ends of lines
input = input.replaceAll(',', '.,');
input = input.split(',');
let out = 0;
let numberMatrix = [...Array(140)].map(e => Array(141));
input.forEach((line, indexL)=>{
	let split = line.split("");
  let constructedNumber = '';
  let indices = [];
  split.forEach((sChar, indexC)=>{
  	if (isNaN(parseInt(sChar))){
			indices.forEach((ind)=>{
      	numberMatrix[ind[0]][ind[1]] = parseInt(constructedNumber);
      });
      indices = [];
      constructedNumber = '';
    }
    else {
    	constructedNumber += sChar;
      indices.push([indexL, indexC])
    }
  });
});
input.forEach((line, indexL)=>{
	let split = line.split("");
  split.forEach((sChar, indexC)=>{
  	if (sChar === '*'){
    	let negL = indexL - 1 >= 0;
      let negC = indexC - 1 >= 0;
      let posL = indexL + 1 < input.length;
      let posC = indexC + 1 < split.length;
      let adjacencyArr = [... negL ? [... negC ? [numberMatrix[indexL-1][indexC-1]] : [], numberMatrix[indexL-1][indexC], ... posC ? [numberMatrix[indexL-1][indexC+1]] : []] : [],
      										... negC ? [numberMatrix[indexL][indexC-1]] : [], ... posC ? [numberMatrix[indexL][indexC+1]] : [],
                          ... posL ? [... negC ? [numberMatrix[indexL+1][indexC-1]] : [], numberMatrix[indexL+1][indexC], ... posC ? [numberMatrix[indexL+1][indexC+1]] : []] : []];
      let gears = [];
      adjacencyArr.forEach((item, index)=>{
      	if (!isNaN(parseInt(item)) && (!gears.includes(item) || adjacencyArr[index-1] !== item)){
        	gears.push(parseInt(item))
        }
      });
      if (gears.length == 2){
      	out += gears[0]*gears[1];
      }
    }
  });
});
console.log(out);

File: 2021/aoc-day3-p2.py
# Put csv input here:
x = ""

y = x.split(",")
c = x.split(",")
b = 0
count = 0
while len(y) > 1 or len(c) > 1:
    count = 0
    for z in y:
        count += int(z[b])
    if count >= len(y)/2:
        y = y if len(y) == 1 else [num for num in y if num[b] == "1"]
    else:
        y = y if len(y) == 1 else [num for num in y if num[b] == "0"]
    count = 0
    for z in c:
        count += int(z[b])
    if count >= len(c)/2:
        c = c if len(c) == 1 else [num for num in c if num[b] == "0"]
    else:
        c = c if len(c) == 1 else [num for num in c if num[b] == "1"]
    b += 1
print("Result: " + str(int(y[0], 2)*int(c[0], 2)))
